plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/Internal.hs

```diff
@@ -562,14 +562,14 @@ evalBuiltinApp
     -> CekValEnv uni fun
     -> BuiltinRuntime (CekValue uni fun)
     -> CekM uni fun s (CekValue uni fun)
-evalBuiltinApp fun term env runtime@(BuiltinRuntime sch x cost) = case sch of
+evalBuiltinApp fun term env runtime@(BuiltinRuntime sch getX cost) = case sch of
     RuntimeSchemeResult -> do
         spendBudgetCek (BBuiltinApp fun) cost
-        let !(errOrRes, logs) = makeKnownRun (Just term) x
+        let !(errOrRes, logs) = runEmitter $ runExceptT getX
         ?cekEmitter logs
         case errOrRes of
-            Left err  -> throwMakeKnownErrorWithCause err
-            Right res -> pure res
+            Left err -> throwReadKnownErrorWithCause $ term <$ err
+            Right x  -> pure x
```

Comment:
Note to self: this should be perfectly fine, but there's a subtlety here: the costing function kinda assumes that the arguments are of particular types, however an argument can be of the wrong type at runtime and the costing function will be applied to it and _something_ will be computed. In the [other](https://github.com/input-output-hk/plutus/pull/4514) PR an argument of the unexpected type is an evaluation error right away. It's probably fine to be a bit sloppy with how costing is managed in case of evaluation failure, but this needs to be documented and discussed.

Reply:
Oh no, it's all wrong. Ok, that's for tomorrow.

Reply:
Right, that is subtle: it fits into the world in which "real" builtins take typed constants and need to be costed appropriately (i.e. taking into account the cases where they're given ill-typed arguments). I think this all works out okay because the runtime-typechecking part is quick and happens first, but it is indeed another dark corner.